core/src/main/java/ai/timefold/solver/core/impl/constructionheuristic/placer/QueuedMultiplePlacer.java

```diff
@@ -0,0 +1,213 @@
+                    // We stop if one of the placement iterators has no next placement
+                    return null;
+                }
+                previousMove = move;
+                cachedMove = CompositeMove.buildMove(move);
+            }
+            return cachedMove;
+        }
+
+        /**
+         * Go through the registered iterators and check for any available moves.
+         *
+         * @param move the move array to be loaded
+         * @param previousMove the previous moves
+         * @return the last index of the iterator that still has available moves; otherwise, the function returns -1
+         *         when all iterators have no more moves available.
+         */
+        private int consumeNextMove(Move<?>[] move, Move<Solution_>[] previousMove) {
+            var index = move.length - 1;
+            // Look for the first iterator that still has available moves to generate
+            while (index >= 0) {
+                var moveIterator = moveIteratorList.get(index);
+                if (moveIterator.hasNext()) {
+                    break;
+                }
+                index--;
+            }
+            if (index < 0) {
+                return -1;
+            }
+            // Copy the previous move until the next one generated
+            System.arraycopy(previousMove, 0, move, 0, index);
+            // Generate and set the new move
+            move[index] = moveIteratorList.get(index).next();
+            return index;
+        }
+
+        /**
+         * Recreate all move iterators starting from #lastValidIteratorIndex.
+         *
+         * @param lastValidIteratorIndex the index of the last iterator that generated a valid move
+         * @param move the move array to be loaded
+         */
+        private boolean recreateNextIterators(int lastValidIteratorIndex, Move<?>[] move) {
+            for (int i = lastValidIteratorIndex + 1; i < move.length; i++) {
+                var placementIterator = queuedPlacerList.get(i).iterator();
+                placementIteratorList.set(i, placementIterator);
+                Move<Solution_> next;
+                if (!placementIterator.hasNext()) {
+                    return false;
+                } else {
+                    var moveIterator = placementIterator.next().iterator();
+                    moveIteratorList.set(i, moveIterator);
+                    next = moveIterator.next();
+                }
+                move[i] = next;
+            }
+            return true;
+        }
+
+        private void clearCache() {
+            this.cachedMove = null;
+        }
+
+        private void reset() {
+            placementIteratorList = new ArrayList<>(queuedPlacerList.size());
+            moveIteratorList = new ArrayList<>(queuedPlacerList.size());
+            for (EntityPlacer<Solution_> placements : queuedPlacerList) {
+                moveIteratorList.add(null);
+                placementIteratorList.add(null);
```

Comment:
This is strange code. If the default value has to be null, why not use an array instead? Or at the very least explain that this is done so that you can do `add(i, ...)` later without getting an exception.